### DISCUSSION TITLE
Is it safe to run JSBSim at a variable timestep?

### DISCUSSION THREAD

**Discussion by aarondewindt** [JSBSim-Team/jsbsim#449] [Q&A]
I want to run JSBSim in realtime with some code in the loop. I tried running it with a fixed timestep, but this didn't work reliably. So I tried adjusting the timestep by calling `FGFDMExec.Setdt(...)` on each iteration based on the realtime clock, simulation time and target timestep and it seems like this has done the trick. 

However I'm not sure whether it's safe to do this, since it may have sideffects depending on what integration scheme is being used in JSBSim. Or maybe there is actual support for this and I need to enable some setting, change integrator, do nothing, etc.

**Comment by seanmcleod** [comment]
Is your issue that there is some amount of jitter between the simulation steps when run in realtime mode? And the amount of jitter was causing an issue with your own code within this loop?

I don't have a definitive answer in terms of the potential side-effects of varying the timestep during a simulation run. There could potentially be some code that caches the timestep value when it's initialized with the assumption that it would remain fixed for the run.

Coincidentally I was looking at the realtime case in JSBSim not that long ago because I noticed some amount of jitter when using JSBSim to update an aircraft pose in Prepar3d when trying to fly in close formation with a Prepar3d controlled aircraft.

JSBSim in realtime mode will basically check for any lag and if it detects any will then run multiple timesteps back to back to try and catch up.

https://github.com/JSBSim-Team/jsbsim/blob/a35be5ce6f7d7ca0ef9a11a62563e54d769bc4f4/src/JSBSim.cpp#L532-L550

**Reply by aarondewindt** [reply]
Thanks, I didn't actually realize there was a realtime mode, I'm linking against the library and interacting with the `FGFDMExec` instance directly.

I don't think the issue I had was related with jitter between JSBSim calls though, it was probably on our side. The difference was quite big, it would lag by about 1sec after just a 5sec simulation. I'll just try implementing it the same way you guys are doing.

**Comment by bcoconni** [comment]
Just for the record, using a variable time step with JSBSim is in most cases a terrible idea. The assumption of a constant time step is used almost everywhere in JSBSim code starting with the Adams-Bashforth integration schemes in `FGPropagate`.  There are also classes such as `FGFilter` which initializes its parameters with the initial time step `dt` and never updates them after (unless one of the filter parameter was defined with a property in which case `FGFilter::CalculateDynamicFilters` is called at each time step - provided that the channel execution rate is set to 1).

So the rule of thumb is: unless you exactly know what you are doing (meaning you are very familiar with JSBSim inner workings), don't use a variable time step with JSBSim. Or even simpler, don't use variable time steps with JSBSim. I, for one, wouldn't wander in such murky waters 😉
